File: bytebrew-cli/src/indexing/embeddings.ts
// Ollama embeddings client
import { IEmbeddingsClient } from '../domain/store.js';
import { getLogger } from '../lib/logger.js';

const DEFAULT_OLLAMA_URL = 'http://localhost:11434';
const DEFAULT_MODEL = 'nomic-embed-text';
const DEFAULT_DIMENSION = 768;
const DEFAULT_TIMEOUT = 60000; // 60 seconds
const DEFAULT_BATCH_SIZE = 50;
const MAX_RETRIES = 3;
const RETRY_DELAY_MS = 1000;
// nomic-embed-text context: 8192 tokens. ~4 chars/token → 32k chars.
// Use conservative limit as safety net when Ollama truncate: true doesn't work.
const MAX_EMBED_CHARS = 28000;

export interface EmbeddingsConfig {
  baseUrl?: string;
  model?: string;
  dimension?: number;
  timeout?: number;
  batchSize?: number;
}

// Ollama /api/embed request (newer endpoint with truncate support)
interface OllamaEmbedRequest {
  model: string;
  input: string | string[];
  truncate?: boolean;
}

// Ollama /api/embed response
interface OllamaEmbedResponse {
  model: string;
  embeddings: number[][];
}

export class EmbeddingsClient implements IEmbeddingsClient {
  private baseUrl: string;
  private model: string;
  private dimension: number;
  private timeout: number;
  private batchSize: number;

  constructor(config: EmbeddingsConfig = {}) {
    this.baseUrl = config.baseUrl || DEFAULT_OLLAMA_URL;
    this.model = config.model || DEFAULT_MODEL;
    this.dimension = config.dimension || DEFAULT_DIMENSION;
    this.timeout = config.timeout || DEFAULT_TIMEOUT;
    this.batchSize = config.batchSize || DEFAULT_BATCH_SIZE;
  }

  async embed(text: string): Promise<number[]> {
    const logger = getLogger();
    let lastError: Error | null = null;

    // Client-side truncation as safety net — some Ollama versions ignore truncate: true
    const truncated = text.length > MAX_EMBED_CHARS ? text.slice(0, MAX_EMBED_CHARS) : text;

    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      try {
        const request: OllamaEmbedRequest = {
          model: this.model,
          input: truncated,
          truncate: true,
        };

        const response = await fetch(`${this.baseUrl}/api/embed`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(request),
          signal: controller.signal,
        });

        if (!response.ok) {
          const body = await response.text();
          const err = new Error(`Ollama API error: status ${response.status}, body: ${body}`);
          // 400 = bad request, retrying won't help
          if (response.status === 400) throw err;
          throw err;
        }

        const data = (await response.json()) as OllamaEmbedResponse;
        if (!data.embeddings || data.embeddings.length === 0) {
          throw new Error('Ollama returned empty embeddings');
        }
        return data.embeddings[0];
      } catch (error) {
        lastError = error as Error;
        clearTimeout(timeoutId);

        // Don't retry client errors (400) — input is bad, retrying won't fix it
        if (lastError.message.includes('status 400')) break;

        if (attempt < MAX_RETRIES) {
          logger.warn(`Embedding attempt ${attempt} failed, retrying...`, {
            error: lastError.message,
          });
          await this.delay(RETRY_DELAY_MS * attempt);
        }
      } finally {
        clearTimeout(timeoutId);
      }
    }

    logger.warn('Embedding failed, skipping chunk', { error: lastError?.message, textLength: text.length });
    throw lastError;
  }

  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  async embedBatch(texts: string[]): Promise<(number[] | null)[]> {
    if (texts.length === 0) {
      return [];
    }

    const logger = getLogger();
    const results: (number[] | null)[] = [];

    for (let i = 0; i < texts.length; i += this.batchSize) {
      const batch = texts.slice(i, i + this.batchSize);

      // Try native batch API first (1 HTTP request for entire batch)
      const batchResult = await this.embedBatchNative(batch);
      if (batchResult) {
        results.push(...batchResult);
        continue;
      }

      // Fallback: embed individually (handles per-item failures)
      logger.debug('Batch embed failed, falling back to individual requests');
      for (const text of batch) {
        try {
          results.push(await this.embed(text));
        } catch {
          results.push(null);
        }
      }
    }

    return results;
  }

  /**
   * Send all texts in one Ollama /api/embed request (input: string[]).
   * Returns null if the batch request itself fails (caller should fallback).
   */
  private async embedBatchNative(texts: string[]): Promise<(number[] | null)[] | null> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout * 2);

    try {
      const truncated = texts.map(t =>
        t.length > MAX_EMBED_CHARS ? t.slice(0, MAX_EMBED_CHARS) : t
      );

      const request: OllamaEmbedRequest = {
        model: this.model,
        input: truncated,
        truncate: true,
      };

      const response = await fetch(`${this.baseUrl}/api/embed`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request),
        signal: controller.signal,
      });

      if (!response.ok) {
        return null; // Let caller fallback to individual
      }

      const data = (await response.json()) as OllamaEmbedResponse;
      if (!data.embeddings || data.embeddings.length !== texts.length) {
        return null;
      }

      return data.embeddings;
    } catch {
      return null;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async ping(): Promise<boolean> {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 5000);

      const response = await fetch(`${this.baseUrl}/api/tags`, {
        method: 'GET',
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      return response.ok;
    } catch {
      return false;
    }
  }

  getDimension(): number {
    return this.dimension;
  }

  getModel(): string {
    return this.model;
  }
}

File: bytebrew/engine/internal/infrastructure/indexing/embeddings.go
package indexing

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"
)

const (
	DefaultOllamaURL = "http://localhost:11434"
	DefaultEmbedModel = "nomic-embed-text"
	DefaultDimension  = 768
	MaxTextLength     = 28000
	EmbedBatchSize    = 50
	maxRetries        = 3
)

// EmbeddingsClient calls the Ollama embeddings API over HTTP.
type EmbeddingsClient struct {
	baseURL   string
	model     string
	dimension int
	client    *http.Client
}

// NewEmbeddingsClient creates a client for the given Ollama instance.
func NewEmbeddingsClient(baseURL, model string, dimension int) *EmbeddingsClient {
	return &EmbeddingsClient{
		baseURL:   baseURL,
		model:     model,
		dimension: dimension,
		client: &http.Client{
			Timeout: 60 * time.Second,
		},
	}
}

// Dimension returns the configured embedding dimension.
func (c *EmbeddingsClient) Dimension() int {
	return c.dimension
}

// Ping checks whether the Ollama server is reachable.
func (c *EmbeddingsClient) Ping(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.baseURL+"/api/tags", nil)
	if err != nil {
		return false
	}
	resp, err := c.client.Do(req)
	if err != nil {
		return false
	}
	resp.Body.Close()
	return resp.StatusCode == http.StatusOK
}

// Embed generates an embedding for a single text.
func (c *EmbeddingsClient) Embed(ctx context.Context, text string) ([]float32, error) {
	results, err := c.EmbedBatch(ctx, []string{text})
	if err != nil {
		return nil, err
	}
	if len(results) == 0 || results[0] == nil {
		return nil, fmt.Errorf("empty embedding result")
	}
	return results[0], nil
}

// EmbedBatch generates embeddings for multiple texts.
// Returns nil entries for texts that failed to embed.
func (c *EmbeddingsClient) EmbedBatch(ctx context.Context, texts []string) ([][]float32, error) {
	if len(texts) == 0 {
		return nil, nil
	}

	// Truncate texts that are too long
	truncated := make([]string, len(texts))
	for i, t := range texts {
		if len(t) > MaxTextLength {
			truncated[i] = t[:MaxTextLength]
		} else {
			truncated[i] = t
		}
	}

	// Try native batch first
	results, err := c.embedBatchNative(ctx, truncated)
	if err == nil {
		return results, nil
	}

	slog.WarnContext(ctx, "native batch embed failed, falling back to individual requests", "error", err)
	return c.embedBatchIndividual(ctx, truncated)
}

// embedRequest is the JSON body for the Ollama embed endpoint.
type embedRequest struct {
	Model    string   `json:"model"`
	Input    []string `json:"input"`
	Truncate bool     `json:"truncate"`
}

// embedResponse is the JSON response from the Ollama embed endpoint.
type embedResponse struct {
	Embeddings [][]float32 `json:"embeddings"`
}

// embedBatchNative sends all texts in a single API call.
func (c *EmbeddingsClient) embedBatchNative(ctx context.Context, texts []string) ([][]float32, error) {
	body, err := json.Marshal(embedRequest{
		Model:    c.model,
		Input:    texts,
		Truncate: true,
	})
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	respBody, err := c.doWithRetry(ctx, body)
	if err != nil {
		return nil, err
	}

	var resp embedResponse
	if err := json.Unmarshal(respBody, &resp); err != nil {
		return nil, fmt.Errorf("unmarshal response: %w", err)
	}

	if len(resp.Embeddings) != len(texts) {
		return nil, fmt.Errorf("expected %d embeddings, got %d", len(texts), len(resp.Embeddings))
	}

	// Convert float64 results to float32 (if needed, Ollama returns float64)
	results := make([][]float32, len(resp.Embeddings))
	for i, emb := range resp.Embeddings {
		results[i] = emb
	}
	return results, nil
}

// embedBatchIndividual sends one request per text as a fallback.
func (c *EmbeddingsClient) embedBatchIndividual(ctx context.Context, texts []string) ([][]float32, error) {
	results := make([][]float32, len(texts))

	for i, text := range texts {
		body, err := json.Marshal(embedRequest{
			Model:    c.model,
			Input:    []string{text},
			Truncate: true,
		})
		if err != nil {
			continue
		}

		respBody, err := c.doWithRetry(ctx, body)
		if err != nil {
			slog.WarnContext(ctx, "embed individual failed", "index", i, "error", err)
			continue
		}

		var resp embedResponse
		if err := json.Unmarshal(respBody, &resp); err != nil {
			continue
		}

		if len(resp.Embeddings) > 0 {
			results[i] = resp.Embeddings[0]
		}
	}

	return results, nil
}

// doWithRetry performs the HTTP request with exponential backoff.
func (c *EmbeddingsClient) doWithRetry(ctx context.Context, body []byte) ([]byte, error) {
	var lastErr error
	for attempt := 0; attempt < maxRetries; attempt++ {
		if attempt > 0 {
			backoff := time.Duration(1<<uint(attempt-1)) * time.Second
			select {
			case <-ctx.Done():
				return nil, ctx.Err()
			case <-time.After(backoff):
			}
		}

		req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+"/api/embed", bytes.NewReader(body))
		if err != nil {
			return nil, fmt.Errorf("create request: %w", err)
		}
		req.Header.Set("Content-Type", "application/json")

		resp, err := c.client.Do(req)
		if err != nil {
			lastErr = fmt.Errorf("http request: %w", err)
			continue
		}

		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()

		if resp.StatusCode == http.StatusBadRequest {
			return nil, fmt.Errorf("bad request (400): %s", string(respBody))
		}

		if resp.StatusCode != http.StatusOK {
			lastErr = fmt.Errorf("status %d: %s", resp.StatusCode, string(respBody))
			continue
		}

		if err != nil {
			lastErr = fmt.Errorf("read response: %w", err)
			continue
		}

		return respBody, nil
	}

	return nil, fmt.Errorf("embed failed after %d attempts: %w", maxRetries, lastErr)
}
